src/ast/codegen_llvm.cpp

```diff
@@ -473,6 +473,58 @@ void CodegenLLVM::visit(Call &call)
     expr_ = buf;
     expr_deleter_ = [this,buf]() { b_.CreateLifetimeEnd(buf); };
   }
+  else if (call.func == "buf")
+  {
+    Value *max_length = b_.getInt64(bpftrace_.strlen_);
+    size_t fixed_buffer_length = bpftrace_.strlen_;
+    Value *length;
+
+    if (call.vargs->size() > 1)
+    {
+      auto &arg = *call.vargs->at(1);
+      arg.accept(*this);
+
+      Value *proposed_length = expr_;
+      Value *cmp = b_.CreateICmp(
+          CmpInst::ICMP_ULE, proposed_length, max_length, "length.cmp");
+      length = b_.CreateSelect(
+          cmp, proposed_length, max_length, "length.select");
+
+      if (arg.is_literal)
+        fixed_buffer_length = static_cast<Integer &>(arg).n;
+    }
+    else
+    {
+      auto &arg = *call.vargs->at(0);
+      fixed_buffer_length = arg.type.pointee_size * arg.type.size;
+      length = b_.getInt8(fixed_buffer_length);
+    }
+
+    auto elements = AsyncEvent::Buf().asLLVMType(b_, fixed_buffer_length);
+    char dynamic_sized_struct_name[30];
+    sprintf(dynamic_sized_struct_name, "buffer_%ld_t", fixed_buffer_length);
```

Comment:
This feels a bit forced, but it's needed so that we don't trip over the caching mechanism for struct types [here](https://github.com/iovisor/bpftrace/blob/e3b9518bb58b785560f84093804d3e7e27533187/src/ast/irbuilderbpf.cpp#L42-L44). The caching assumes that structs with the same name will have the same layout and sizes, and that wasn't true in my earlier implementation.

Reply:
See comment above. I think we can avoid this.